File: coding/src/com/algos/core/li08_queue/revised/L232_ImplementQueueUsingStacks.java
package  com.algos.core.li08_queue.revised;

import java.util.Stack;

public class L232_ImplementQueueUsingStacks {

	public static void main(String[] args) {
		L232_ImplementQueueUsingStacks obj = new L232_ImplementQueueUsingStacks();
		
		obj.push(1);
		obj.push(2);
		obj.push(3);
		System.out.println(obj.peek());
		System.out.println(obj.pop());
		System.out.println(obj.empty());
	}

	// Problem: https://leetcode.com/problems/implement-queue-using-stacks/
	// Idea: Use two stacks. s1->s2, s1.push(x), s2->s1.
    // Alternative to achieve O(1) armotized complexity. Use two stacks. input and output. Push elements to input stack and pop elements from output stack.
	// If output stack is empty move all the elements from input stack to output stack.
	Stack<Integer> stack1;
    Stack<Integer> stack2;
    public L232_ImplementQueueUsingStacks() {
        stack1 = new Stack<>();
        stack2 = new Stack<>();
    }
    
    public void push(int x) {
        while(!stack1.isEmpty()) stack2.push(stack1.pop());
        stack1.push(x);
        while(!stack2.isEmpty()) stack1.push(stack2.pop());
    }
    
    public int pop() {
        return stack1.pop();
    }
    
    public int peek() {
        return stack1.peek();
    }
    
    public boolean empty() {
        return stack1.isEmpty();
    }
}

File: dsa/concepts/08_queues/python/l232_implement_queue_using_stacks.py
"""
Problem: https://leetcode.com/problems/implement-queue-using-stacks/
Idea:
Use two stacks.
Approach shown here:
- Move all elements from stack1 to stack2
- Push new element to stack1
- Move everything back from stack2 to stack1
This ensures front of queue is always on top of stack1.

Alternative (Optimal):
Use input and output stacks to achieve O(1) amortized time.
"""

class MyQueue:
    def __init__(self):
        self.stack1 = []
        self.stack2 = []

    def push(self, x: int) -> None:
        # Move all elements to stack2
        while self.stack1:
            self.stack2.append(self.stack1.pop())

        # Push new element
        self.stack1.append(x)

        # Move back to stack1
        while self.stack2:
            self.stack1.append(self.stack2.pop())

    def pop(self) -> int:
        return self.stack1.pop()

    def peek(self) -> int:
        return self.stack1[-1]

    def empty(self) -> bool:
        return not self.stack1


# Example usage
if __name__ == "__main__":
    q = MyQueue()

    q.push(1)
    q.push(2)
    q.push(3)
    print(q.peek())   # 1
    print(q.pop())    # 1
    print(q.empty())  # False
